### ISSUE TITLE
Unable to retrieve status via proxy with non-default http port

### ISSUE THREAD

**Issue by alanbirtles** [erikdoe/ccmenu#77]
I need to access a Jenkins server through a http proxy, I've set the server url to:
` http://username@1.2.3.4:8080/cc.xml `
Looking at wireshark the request being sent to the proxy is
```
GET http://1.2.3.4/cc.xml HTTP/1.1
Host: 1.2.3.4:8080
Accept: */*
Proxy-Connection: keep-alive
Cookie: JSESSIONID.73a11962=2lnyztaooucjb2rpazzpq7n5
User-Agent: CCMenu/1121.2 CFNetwork/807.2.14 Darwin/16.3.0 (x86_64)
Accept-Language: en-gb
Accept-Encoding: gzip, deflate
Connection: keep-alive
```
the port number should be in the GET line in addition to the Host line.
I'm using version 12.1

**Comment by erikdoe** [comment]
I'm sorry but the output from Wireshark doesn't make sense. How did you get it? I'm asking because in the HTTP protocol (version 1.0 and 1.1) what follows the `GET` verb is the path only. The scheme, hostname, and port number do not belong there, and are not sent. Is this maybe a "helpful" output option in Wireshark?

In any case, the correct HTTP request should look like this:

    GET /cc.xml HTTP/1.1
    Host: 1.2.3.4:8080

And I'm sure that Apple's HTTP libraries, which CCMenu uses, implement HTTP correctly. If you specify a port, then CCMenu uses it. That said, maybe the proxy is misconfigured and can't handle it?

**Comment by alanbirtles** [comment]
The wireshark trace is the communication between ccmenu and the proxy, it is mostly correct but should be:
```
GET http://1.2.3.4:8080/cc.xml HTTP/1.1
Host: 1.2.3.4:8080
Accept: */*
Proxy-Connection: keep-alive
Cookie: JSESSIONID.73a11962=2lnyztaooucjb2rpazzpq7n5
User-Agent: CCMenu/1121.2 CFNetwork/807.2.14 Darwin/16.3.0 (x86_64)
Accept-Language: en-gb
Accept-Encoding: gzip, deflate
Connection: keep-alive
```

**Comment by alanbirtles** [comment]
this is the request sent by safari to the proxy for the same URL
```
GET http://1.2.3.4:8080/cc.xml HTTP/1.1
Host: 1.2.3.4:8080
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Proxy-Connection: keep-alive
Upgrade-Insecure-Requests: 1
Cookie: JSESSIONID.73a11962=7cj4g47fmhulj285208nw22j
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_2) AppleWebKit/602.3.12 (KHTML, like Gecko) Version/10.0.2 Safari/602.3.12
Accept-Language: en-gb
Accept-Encoding: gzip, deflate
Connection: keep-alive
```

**Comment by alanbirtles** [comment]
interestingly the correct request is sent when no authentication is specified

**Comment by alanbirtles** [comment]
not specifying the credentials in the application and hard coding them into the addBasicAuthToRequest method seems to fix the issue, can't work out where its going wrong in the first place though.
Let me know if you need any help reproducing the issue. I think if you set your system proxy settings to point at a local http server you should be able to reproduce the issue.

**Comment by alanbirtles** [comment]
removing the username from the url before passing it to NSMutableURLRequest seems to fix the issue (presumably this is a bug in NSMutableURLRequest?).

e.g. change the beginning of CCMConnection.createRequest to
```
    NSString* tempUrlString = [feedURL absoluteString];
    tempUrlString = [tempUrlString stringByReplacingCredentials:@""];
    NSURL* tempURL = [NSURL URLWithString:tempUrlString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:tempURL cachePolicy:NSURLRequestReloadIgnoringCacheData 
```

**Comment by erikdoe** [comment]
First off, I don't know what Wireshark does and why it does it, but the HTTP standard really does not allow the scheme, hostname, and port to be included after the `GET` verb. I'm not saying that Wireshark doesn't print the text you've pasted (haven't used Wireshark in a while) but I am sure that this is not the text that is actually sent over the wire. 

Have a look at the description of HTTP on Wikipedia, the sections in red: https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#/media/File:Http_request_telnet_ubuntu.png

The *path* following the `GET` verb has been in HTTP since 1.0, and it simply didn't make sense to include the hostname and port there, because the text is sent over a TCP connection that is *already established* to that host and port. The `Host` header was added in 1.1 to allow for virtual hosts to be hosted on the same IP address.  Why would the standard allow the inclusion of a hostname after the `GET` verb *and* in the `Host` header? And what would a server implementation do if these don't correspond?

Either way, I have just re-tested this manually. I've launched the fake server in `Tools/fakeserver.rb` as this does run on a non-standard port and offers a password protected endpoint. Basically I've added the following feed URL to CCMenu: `http://dev:rosebud@localhost:4567/protected/cctray.xml`

It works without a flaw. To me, this shows that CCMenu in general and `NSMutableURLRequest` in particular can handle URLs that include both, user credentials and port numbers.

Are you absolutely certain that your proxy isn't causing the issue? Could you provide a piece of code that demonstrates how `NSMutableURLRequest` or `NSURL` are buggy?

**Comment by alanbirtles** [comment]
yes I'm certain, I agree normal HTTP get verbs only include the path but when using a http proxy server is involved the client passes the full url in the get line to the proxy and the proxy then connects to the target server and sends a normal http request containing just the path in the get line.
The traces I've posted are the messages sent between ccmenu and the proxy.
what I meant by setting up a server was to set the server as a http proxy in your system settings so you could see this behaviour for yourself.

**Comment by alanbirtles** [comment]
see http://stackoverflow.com/questions/7577917/how-does-a-http-proxy-utilize-the-http-protocol-a-proxy-rfc for example

**Comment by erikdoe** [comment]
Okay, now I understand. When a proxy is specified, the requesting client writes the HTTP request so that it includes the absolute form, and not just the path. 

Using a proxy I can now reproduce the behaviour you described; and I can see it without Wireshark. So, it looks like I was over-confident in the quality of `NSURLRequest` and I concur that this really does look like a bug in it.

I can also confirm that removing the credentials from the URL passed to `NSURLRequest` results in the correct HTTP request being written. 

At the moment I can't think of a reason not to apply this fix, but I'm not sure, just having demonstrated that I don't know HTTP as well as I thought. ;-) Could there be a reason why the username would have to be passed to `NSURLRequest`? (Under normal circumstances, only the username is in the feed URL because the password is stored in keychain and set differently.)

**Comment by alanbirtles** [comment]
I don't think so, NSURLRequest doesn't seem to do much with the username in the URL other than removing it before creating the HTTP request (I guess it's this removal that is broken?)

**Comment by erikdoe** [comment]
Released in CCMenu 13.0.
